File: leetcode/328.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function oddEvenList(head: ListNode | null): ListNode | null {
    let cur = head;
    let oddHead = null;
    let oddTail = null;
    while(cur!=null && cur.next != null && cur.next.next != null){
        // console.log(cur.val);
        if(oddHead == null){
            oddHead = cur.next;
            oddTail = cur.next;
        }
        else{
            oddTail.next = cur.next;
        }
        oddTail = cur.next;
        cur.next = cur.next.next;
        cur = cur.next;
    }
    if(cur == null){
        return head;
    }
    else if(cur.next == null){
        cur.next = oddHead;
        if(oddTail){
            oddTail.next = null;
        }
    }
    else{
        if(!oddHead){
            oddHead = cur.next;
            oddTail = cur.next;
        }
        else{
            oddTail.next = cur.next;
            oddTail = cur.next;
        }
        oddTail.next = null;
        cur.next = oddHead;
    }
    return head;
};

File: leetcode/328.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode OddEvenList(ListNode head) {
        if(head == null){
            return null;
        }
        if(head.next == null){
            return head;
        }
        ListNode oddHead = head;
        ListNode evenHead = head.next;
        ListNode oddTail = head;
        ListNode evenTail = head.next;
        ListNode cur = evenTail.next;
        while(cur != null && cur.next != null){
            oddTail.next = cur;
            oddTail = cur;
            evenTail.next = cur.next;
            evenTail = cur.next;
            cur = cur.next.next;
        }
        if(cur != null){
            oddTail.next = cur;
            oddTail = cur;
        }
        oddTail.next = evenHead;
        evenTail.next = null;
        return oddHead;
    }
}
